allennlp/training/trainer.py

```diff
@@ -769,12 +768,37 @@ def train(self) -> Dict[str, Any]:
                 # No validation set, so just assume it's the best so far.
                 is_best_so_far = True
                 val_metrics = {}
-                best_epoch_val_metrics = {}
                 this_epoch_val_metric = None
 
             self._metrics_to_tensorboard(epoch, train_metrics, val_metrics=val_metrics)
             self._metrics_to_console(train_metrics, val_metrics)
 
+            # Create overall metrics dict
+            training_elapsed_time = time.time() - training_start_time
+            metrics["training_duration"] = time.strftime("%H:%M:%S", time.gmtime(training_elapsed_time))
+            metrics["training_start_epoch"] = epoch_counter,
+            metrics["training_epochs"] = epochs_trained
+            metrics["epoch"] = epoch
+
+            for key, value in train_metrics.items():
+                metrics["training_" + key] = value
+            for key, value in val_metrics.items():
+                metrics["validation_" + key] = value
+
+            if is_best_so_far:
+                metrics['best_epoch'] = epoch
+                for key, value in val_metrics.items():
+                    metrics["best_validation_" + key] = value
+
+            if self._serialization_dir:
+                dump_metrics(os.path.join(self._serialization_dir, f'metrics_epoch_{epoch}.json'), metrics)
+
+            for index, param_group in enumerate(self._optimizer.param_groups):
```

Comment:
Is this from a merge or something? It seems unrelated

Reply:
oops, yeah I had a bad merge, let me fix this